### PR TITLE
Add support for resumable signed url generation

### PR DESCRIPTION
**Pull request by ThatsAMorais**
Modified `generate_signed_url` to support a similar feature in
gsutil for generating signed URLs that enable clients to initiate
an upload, particularly in their region rather than in the
server's.

### CODE REVIEW

**Comment by ThatsAMorais** [comment]
I signed it!

**Comment by chemelnucfin** [comment]
Hello, it may be possible that you used another e-mail to sign the CLA and to use for your github commit.  Could you please try ensuring the e-mail you used are the same for both?

Thank you.

**Comment by ThatsAMorais** [comment]
I think you're right. I resigned it with my personal email. I was probably logged in as a different gmail account in the browser where I signed it.

**Comment by ThatsAMorais** [comment]
I signed it

**Comment by ThatsAMorais** [comment]
I am not sure where the Signed URL documentation is hosted or I would update that with this flow. Can anyone point me to it if its hosted in a public repo on Github?

**Comment by ThatsAMorais** [comment]
Okay, finally got the new test passing. Ready for review.

**Comment by jonparrott** [comment]
Hey @ThatsAMorais! I just wanted to let you know that I'm traveling and won't have a chance to review this until late next week. Thank you for doing this!

File: storage/google/cloud/storage/_signing.py [hunk 1]
```diff
@@ -143,7 +143,11 @@ def generate_signed_url(credentials, resource, expiration,
 
     :type method: str
     :param method: The HTTP verb that will be used when requesting the URL.
-                   Defaults to ``'GET'``.
+                   Defaults to ``'GET'``. If method is ``'RESUMABLE'`` then the
+                   signature will additionally contain the `x-goog-resumable`
+                   header, and the method changed to POST. See the signed URL
+                   docs regarding this flow.
```
Comment (jonparrott) [inline storage/google/cloud/storage/_signing.py]: Can we add an actual link to the signed url documentation?

File: storage/google/cloud/storage/_signing.py [hunk 2]
```diff
@@ -172,13 +176,20 @@ def generate_signed_url(credentials, resource, expiration,
     """
     expiration = get_expiration_seconds(expiration)
 
+    if method == "RESUMABLE":
```
Comment (jonparrott) [inline storage/google/cloud/storage/_signing.py]: nit: use single quotes

File: storage/tests/unit/test_blob.py [hunk 3]
```diff
@@ -312,6 +313,32 @@ def test_generate_signed_url_w_method_arg(self):
...
+                return_value={
+                    'GoogleAccessId': 'service-account-name',
+                    'Expires': 12345,
+                    'Signature': 'signed-data',
+                })
+    def test_generate_resumable_signed_url(self, mock_get_signed_query_params):
+        """
+        Verify correct behavior of resumable upload URL generation
+        """
+        from google.cloud.storage._signing import generate_signed_url
```
Comment (jonparrott) [inline storage/tests/unit/test_blob.py]: nit: empty newline between imports and code.

File: storage/tests/unit/test_blob.py [hunk 4]
```diff
@@ -312,6 +313,32 @@ def test_generate_signed_url_w_method_arg(self):
...
+        expiry = datetime.datetime.utcnow() + datetime.timedelta(hours=1)
+        micros = \
```
Comment (jonparrott) [inline storage/tests/unit/test_blob.py]: I thought we had a helper for this.

Comment (ThatsAMorais) [reply storage/tests/unit/test_blob.py]: Done.

File: storage/tests/unit/test_blob.py (hunk 3 above)
Comment (ThatsAMorais) [reply storage/tests/unit/test_blob.py]: Done.

File: storage/google/cloud/storage/_signing.py (hunk 2 above)
Comment (ThatsAMorais) [reply storage/google/cloud/storage/_signing.py]: Done.

File: storage/google/cloud/storage/_signing.py (hunk 1 above)
Comment (ThatsAMorais) [reply storage/google/cloud/storage/_signing.py]: Done.

File: storage/tests/unit/test_blob.py [hunk 5]
```diff
@@ -312,6 +313,33 @@ def test_generate_signed_url_w_method_arg(self):
...
+                    'Signature': 'signed-data',
+                })
+    def test_generate_resumable_signed_url(self, mock_get_signed_query_params):
+        """
+        Verify correct behavior of resumable upload URL generation
+        """
+        from google.cloud.storage._signing import (
+            generate_signed_url, get_expiration_seconds
+        )
+
```
Comment (chemelnucfin) [inline storage/tests/unit/test_blob.py]: Please have only one input per line.

Comment (ThatsAMorais) [reply storage/tests/unit/test_blob.py]: Done.

**Comment by ThatsAMorais** [comment]
Sorry, guys, gonna take me a minute to figure out why the branch coverage dropped so low. It does not appear to be from my changes.

**Comment by jonparrott** [comment]
It's unrelated, we need to fix it in master. If you just want to make sure that your changes are covered we should be able to merge.

**Comment by ThatsAMorais** [comment]
👍 None of my files are listed here:
```
Branch BrPart  Cover   Missing
-------------------------------------------------------------------------------------------------------------------
google/cloud/language_v1/__init__.py                                        8      0      0      0   100%
google/cloud/language_v1/gapic/__init__.py                                  0      0      0      0   100%
google/cloud/language_v1/gapic/enums.py                                   206      0      0      0   100%
google/cloud/language_v1/gapic/language_service_client.py                  49      2      6      3    91%   84, 89, 83->84, 88->89, 97->101
google/cloud/language_v1/gapic/language_service_client_config.py            1      0      0      0   100%
google/cloud/language_v1/proto/__init__.py                                  0      0      0      0   100%
google/cloud/language_v1/proto/language_service_pb2.py                    309     45      2      0    86%   2831-2833, 2840-2842, 2848-2850, 2857-2859, 2864-2866, 2872-2874, 2878-2912, 2927, 2933, 2938, 2944, 2948, 2953, 2968, 2975, 2981, 2988, 2993, 2999, 3009-3034, 3043-3070
google/cloud/language_v1/types.py                                          13      0      4      0   100%
google/cloud/language_v1beta2/__init__.py                                   8      0      0      0   100%
google/cloud/language_v1beta2/gapic/__init__.py                             0      0      0      0   100%
google/cloud/language_v1beta2/gapic/enums.py                              206      0      0      0   100%
google/cloud/language_v1beta2/gapic/language_service_client.py             49      2      6      3    91%   84, 89, 83->84, 88->89, 97->101
google/cloud/language_v1beta2/gapic/language_service_client_config.py       1      0      0      0   100%
google/cloud/language_v1beta2/proto/__init__.py                             0      0      0      0   100%
google/cloud/language_v1beta2/proto/language_service_pb2.py               312     45      2      0    86%   2835-2837, 2844-2846, 2852-2854, 2861-2863, 2868-2870, 2876-2878, 2882-2916, 2931, 2937, 2942, 2948, 2952, 2957, 2972, 2979, 2985, 2992, 2997, 3003, 3013-3038, 3047-3074
google/cloud/language_v1beta2/types.py                                     18      0      4      0   100%
-------------------------------------------------------------------------------------------------------------------
TOTAL                                                                    1180     94     24      6    92%
```

**Comment by jonparrott** [comment]
Verified changes locally:

```
nox > Running session unit(py='2.7')
nox > /Users/jonwayne/.pyenv/versions/2.7.13/bin/python2.7 -m virtualenv /Users/jonwayne/workspace/google-cloud-python/storage/.nox/unit-2-7 -p python2.7
nox > chdir /Users/jonwayne/workspace/google-cloud-python/storage
nox > pip install --upgrade mock pytest pytest-cov ../api_core ../core
nox > pip install --upgrade -e .
nox > py.test --quiet --cov=google.cloud.storage --cov=tests.unit --cov-append --cov-config=.coveragerc --cov-report= --cov-fail-under=97 tests/unit --cov-report=term
............................................................................................................................................................................................................................................ [ 55%]
...........................................................................................................................................................................................                                                  [100%]

---------- coverage: platform darwin, python 2.7.13-final-0 ----------
Name                                   Stmts   Miss Branch BrPart  Cover   Missing
----------------------------------------------------------------------------------
google/cloud/storage/__init__.py           8      0      0      0   100%
google/cloud/storage/_helpers.py          68      0     16      0   100%
google/cloud/storage/_http.py              9      0      0      0   100%
google/cloud/storage/_signing.py          41      0     16      0   100%
google/cloud/storage/acl.py              169      0     40      0   100%
google/cloud/storage/batch.py            127      0     26      0   100%
google/cloud/storage/blob.py             367      0     96      0   100%
google/cloud/storage/bucket.py           303      0     80      0   100%
google/cloud/storage/client.py            79      0     20      0   100%
google/cloud/storage/iam.py               23      0      0      0   100%
google/cloud/storage/notification.py     116      0     32      0   100%
tests/unit/__init__.py                     0      0      0      0   100%
tests/unit/test__helpers.py              212      0      0      0   100%
tests/unit/test__http.py                  40      0      0      0   100%
tests/unit/test__signing.py              123      0     10      0   100%
tests/unit/test_acl.py                   662      0      2      0   100%
tests/unit/test_batch.py                 386      0      2      0   100%
tests/unit/test_blob.py                 1719      0     68      0   100%
tests/unit/test_bucket.py               1274      0     28      0   100%
tests/unit/test_client.py                333      0      0      0   100%
tests/unit/test_iam.py                     5      0      0      0   100%
tests/unit/test_notification.py          307      0      2      0   100%
----------------------------------------------------------------------------------
TOTAL                                   6371      0    438      0   100%

Required test coverage of 97% reached. Total coverage: 100.00%
423 passed in 3.31 seconds
nox > Session unit(py='2.7') was successful.
```

(Same for 3.6). #4843 will resolve the unrelated failures. Merging.

**Comment by jonparrott** [comment]
Thank you @ThatsAMorais for doing this!

**Comment by ThatsAMorais** [comment]
Thanks, @jonparrott! I never looked into how to run the tests locally, so thanks for doing that.

**Pull request merged**